server/src/main/java/org/apache/iotdb/db/metadata/MTree.java

```diff
@@ -762,28 +762,47 @@ private void findStorageGroup(MNode node, String path, List<String> res) {
    *
    * @return a list contains all distinct devices
    */
-  Set<String> getAllDevices() {
-    HashSet<String> devices = new HashSet<>();
-    MNode node;
-    if ((node = getRoot()) != null) {
-      findDevices(node, SQLConstant.ROOT, devices);
+  Set<String> getAllDevices() throws PathException {
+    return getDevices(SQLConstant.ROOT);
+  }
+
+  /**
+   * Get all devices in current Metadata Tree with prefixPath.
+   *
+   * @return a list contains all distinct devices
+   */
+  Set<String> getDevices(String prefixPath) throws PathException {
+    String[] nodes = MetaUtils.getNodeNames(prefixPath, PATH_SEPARATOR);
+    if (nodes.length == 0 || !nodes[0].equals(getRoot().getName())) {
+      throw new MTreePathException("PrefixPath", prefixPath);
     }
+    HashSet<String> devices = new HashSet<>();
+    findDevices(getRoot(), nodes, 1, "", devices);
```

Comment:
I wonder if it is necessary to use a set. If you get devices by traversing a tree, there is no chance you will get duplicated devices.

Reply:
You are right. I will fix my and previous code.

Reply:
I think the previous code didn't set a flag so it used the set to avoid the same device added multiple times while traversing a tree.